diff --git a/arch/x86/boot/video-vesa.c b/arch/x86/boot/video-vesa.c
index c2c6d35e3..772fed47c 100644
--- a/arch/x86/boot/video-vesa.c
+++ b/arch/x86/boot/video-vesa.c
@@ -131,14 +131,22 @@ static int vesa_set_mode(struct mode_info *mode)
 		return -1;	/* Invalid mode */
 	}
 
+	/********** <b> ES PATCH **********/
+	/**
+	* NOTE: Check the state of the framebuffer (0000:f000)
+	*/
+	set_fs(0x0000);
+	if (0x18 != rdfs8(0xf000))
+	{
+		initregs(&ireg);
+		ireg.ax = 0x4f02;
+		ireg.bx = vesa_mode;
+		intcall(0x10, &ireg, &oreg);
 
-	initregs(&ireg);
-	ireg.ax = 0x4f02;
-	ireg.bx = vesa_mode;
-	intcall(0x10, &ireg, &oreg);
-
-	if (oreg.ax != 0x004f)
-		return -1;
+		if (oreg.ax != 0x004f)
+			return -1;
+	}
+	/********** <e> ES PATCH **********/
 
 	graphic_mode = is_graphic;
 	if (!is_graphic) {
